internal/mtail/unix_socket_export_integration_test.go

```diff
@@ -0,0 +1,185 @@
+// the server.  It returns the server, or any errors the new server creates.
+func makeServer(tb testing.TB, pollInterval time.Duration, enableFsNotify bool, options ...func(*mtail.Server) error) (*mtail.Server, error) {
+	tb.Helper()
+	w, err := watcher.NewLogWatcher(pollInterval, enableFsNotify)
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	return mtail.New(metrics.NewStore(), w, options...)
+}
+
+
+// startUNIXSocketServer creates a new Server serving through a UNIX
+// socket and starts it running. It returns the server, and a cleanup function.
+func startUNIXSocketServer(tb testing.TB, pollInterval time.Duration, enableFsNotify bool, options ...func(*mtail.Server) error) (*mtail.Server, func()) {
+	tb.Helper()
+
+	unixSocket := "/var/run/mtail_test.socket"
+	options = append(options, mtail.BindUnixSocket("/var/run/mtail_test.socket"))
+
+	m, err := makeServer(tb, pollInterval, enableFsNotify, options...)
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	errc := make(chan error, 1)
+	go func() {
+		err := m.Run()
+		errc <- err
+	}()
+
+	glog.Infof("check that server is listening")
+
+	addr, err := net.ResolveUnixAddr("unix", unixSocket)
+
+	if err != nil {
+		tb.Fatal(err)
+	}
+	_, err = net.DialUnix("unix", nil, addr)
+
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	return m, func() {
+		err := m.Close()
+		if err != nil {
+			tb.Fatal(err)
+		}
+		if err != nil {
+			tb.Fatal(err)
+		}
+		select {
+		case err = <-errc:
+		case <-time.After(5 * time.Second):
+			buf := make([]byte, 1<<16)
+			n := runtime.Stack(buf, true)
+			fmt.Fprintf(os.Stderr, "%s", buf[0:n])
+			tb.Fatal("timeout waiting for shutdown")
+		}
+
+		if err != nil {
+			tb.Fatal(err)
+		}
+	}
+}
+
+
+// getMetricFromUNIXSocket fetches the name metrics from the Server serving through
+// the UNIX socket sockPath, and returns the value of one named name.  Callers are
+// responsible for type assertions on the returned value.
+func getMetricFromUNIXSocket(tb testing.TB, sockPath, name string) interface{} {
+	uri := "http://unix/debug/vars"
+	client := &http.Client{
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial("unix", sockPath)
+			},
+		},
+		Timeout: 5 * time.Second,
+	}
+	resp, err := client.Get(uri)
+	if err != nil {
+		tb.Fatal(err)
+	}
+	buf := new(bytes.Buffer)
+	n, err := buf.ReadFrom(resp.Body)
+	testutil.FatalIfErr(tb, err)
+	glog.V(2).Infof("TestGetMetric: http client read %d bytes from debug/vars", n)
+	var r map[string]interface{}
+	if err := json.Unmarshal(buf.Bytes(), &r); err != nil {
+		tb.Fatalf("%s: body was %s", err, buf.String())
+	}
+	glog.Infof("TestGetMetric: returned value for %s: %v", name, r[name])
+	return r[name]
+}
+
+
+// expectMetricDelta checks to see if the difference between a and b is want;
+// it assumes both values are float64s that came from a TestGetMetric.
+func expectMetricDelta(tb testing.TB, a, b interface{}, want float64) {
```

Comment:
I think you can reuse testing's ExpectMetricDelta instead of copying it here?

Reply:
Initially I tried to do it like you said, by using `mtail.ExpectMetricDelta` and `mtail.TestMakeServer` but when I tested it with the command `go test unix_socket_export_integration_test.go`, the methods were not defined. Perhaps I'm doing it the wrong way (I never used the Go test mechanism before and the Makefile targets don't execute the tests from the `mtail_test` package).